test/system/src/test/java/io/pravega/test/system/AbstractFailoverTests.java

```diff
@@ -97,22 +94,90 @@
         final CompletableFuture<Void> newWritersComplete = new CompletableFuture<>();
         final CompletableFuture<Void> readersComplete = new CompletableFuture<>();
         final List<CompletableFuture<Void>> txnStatusFutureList = synchronizedList(new ArrayList<>());
-        final AtomicBoolean txnWrite = new AtomicBoolean(false);
+
+        final ConcurrentSet<UUID> committingTxn = new ConcurrentSet<>();
+        final ConcurrentSet<UUID> abortedTxn = new ConcurrentSet<>();
+        final AtomicLong eventData = new AtomicLong();
+        final boolean txnWrite;
+
+        private final ConcurrentHashMap<Long, Integer> eventMap = new ConcurrentHashMap<>();
+
+        TestState(boolean txnWrite) {
+            this.txnWrite = txnWrite;
+        }
+
+        void eventWritten(Long event) {
+            eventMap.putIfAbsent(event, 0);
+        }
+
+        void eventRead(Long event) {
+            eventMap.compute(event, (x, y) -> {
+                if (y == null) {
+                    return 1;
```

Comment:
If y is null it implies that an event that was not written was read, right?
We should log this as an error.

Reply:
No before the write ack comes we could have read the event back..